internal/backends/nodejs/nodejs.go

```diff
@@ -309,6 +309,67 @@ var NodejsYarnBackend = api.LanguageBackend{
 	Guess:        nodejsGuess,
 }
 
+var NodejsPNPMBackend = api.LanguageBackend{
+	Name:             "nodejs-pnpm",
+	Specfile:         "package.json",
+	Lockfile:         "pnpm-lock.yaml",
+	FilenamePatterns: nodejsPatterns,
+	Quirks: api.QuirksAddRemoveAlsoLocks |
+		api.QuirksAddRemoveAlsoInstalls |
+		api.QuirksLockAlsoInstalls,
+	GetPackageDir: func() string {
+		return "node_modules"
+	},
+	Search: nodejsSearch,
+	Info:   nodejsInfo,
+	Add: func(pkgs map[api.PkgName]api.PkgSpec, projectName string) {
+		if !util.Exists("package.json") {
+			util.RunCmd([]string{"npm", "init", "-y"})
+		}
+		cmd := []string{"pnpm", "add"}
+		for name, spec := range pkgs {
+			arg := string(name)
+			if spec != "" {
+				arg += "@" + string(spec)
+			}
+			cmd = append(cmd, arg)
+		}
+		util.RunCmd(cmd)
+	},
+	Remove: func(pkgs map[api.PkgName]bool) {
+		cmd := []string{"pnpm", "remove"}
+		for name := range pkgs {
+			cmd = append(cmd, string(name))
+		}
+		util.RunCmd(cmd)
+	},
+	Lock: func() {
+		util.RunCmd([]string{"pnpm", "install"})
+	},
+	Install: func() {
+		util.RunCmd([]string{"pnpm", "install"})
+	},
+	ListSpecfile: nodejsListSpecfile,
+	ListLockfile: func() map[api.PkgName]api.PkgVersion {
+		contentsB, err := ioutil.ReadFile("pnpm-lock.yaml")
+		if err != nil {
+			util.Die("pnpm-lock.yaml: %s", err)
+		}
+		contents := string(contentsB)
+		r := regexp.MustCompile(`(?m)^"?([^@ \n]+).+:\n  version "(.+)"`)
+		pkgs := map[api.PkgName]api.PkgVersion{}
+		for _, match := range r.FindAllStringSubmatch(contents, -1) {
+			name := api.PkgName(match[1])
+			version := api.PkgVersion(match[2])
+			pkgs[name] = version
+		}
+		return pkgs
```

Comment:
Have you considered using `pnpm list --json --depth 0`, instead? I feel like reading lock files and parsing them is error prone and bordering the internals of package manager

Reply:
I guess this is what we do for all other, cc @cdmistman we should consider using commands provided by package managers if they provide a machine readable format

Reply:
agreed that we shouldn't rely on lockfile. if we do, then it's definitely preferred to use the appropriate de/serialization for it - for example, this code shouldn't use a regex, it should instead deserialize the lockfile and read the entries via the deserialized map.

i'm not convinced that shelling out to `pnpm` is the best solution here. it has to load up node, which is big and heavy - we're already concerned about performance in UPM. I think it's better to deserialize; if we're concerned about compatibility, we can error on lockfile versions that are a major version higher than expected.

Reply:
No, I didn't consider the other package manager commands as I tried to do everything in a similar way to how the system for yarn was implemented (hence the regex) and I'm also not very familiar with Go.

Reply:
no worries! i am gonna shell out to pnpm after all - too much effort to deserialize lmao